lib/proxy/peer/quicserver.go

```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
-	panic("QUIC proxy peering is not implemented")
+	s.mu.Lock()
+	if s.closed {
+		s.mu.Unlock()
+		return trace.Wrap(quic.ErrServerClosed)
+	}
+	s.wg.Add(1)
+	defer s.wg.Done()
+	s.mu.Unlock()
+
+	lis, err := t.ListenEarly(s.tlsConfig, s.quicConfig)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer lis.Close()
+	defer context.AfterFunc(s.serveCtx, func() { _ = lis.Close() })()
+
+	for {
+		// the listener will be closed when serveCtx is done, but Accept will
+		// return any queued connection before erroring out with a
+		// [quic.ErrServerClosed]
+		c, err := lis.Accept(context.Background())
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		s.wg.Add(1)
+		go s.handleConn(c)
+	}
+}
+
+func (s *QUICServer) handleConn(c quic.EarlyConnection) {
+	defer s.wg.Done()
+
+	log := s.log.With(
+		"remote_addr", c.RemoteAddr().String(),
+		"internal_id", uuid.NewString(),
+	)
+	state := c.ConnectionState()
+	log.InfoContext(s.serveCtx,
+		"Handling new peer connection.",
+		"gso", state.GSO,
+		"used_0rtt", state.Used0RTT,
+	)
+	defer func() {
+		log.DebugContext(c.Context(),
+			"Peer connection closed.",
+			"error", context.Cause(c.Context()),
+		)
+	}()
+
+	defer c.CloseWithError(0, "")
+	defer context.AfterFunc(s.runCtx, func() { _ = c.CloseWithError(0, "") })()
+
+	for {
+		// TODO(espadolini): stop accepting new streams once s.serveCtx is
+		// canceled, once quic-go gains the ability to change the amount of
+		// available streams during a connection (so we can set it to 0)
+		st, err := c.AcceptStream(context.Background())
+		if err != nil {
+			log.DebugContext(c.Context(), "Got an error accepting a stream.", "error", err)
+			return
+		}
+
+		s.wg.Add(1)
+		go s.handleStream(st, c, log)
+	}
+}
+
+func (s *QUICServer) handleStream(st quic.Stream, c quic.EarlyConnection, log *slog.Logger) {
+	defer s.wg.Done()
+
+	log = log.With("stream_id", st.StreamID())
+	defer log.DebugContext(c.Context(), "Done handling stream.")
+
+	defer st.CancelRead(0)
+	defer st.CancelWrite(0)
+
+	log.DebugContext(c.Context(), "Handling stream.")
+
+	sendErr := func(toSend error) {
+		st.CancelRead(0)
+		defer st.CancelWrite(0)
+		errBuf, err := proto.Marshal(&quicpeeringv1a.DialResponse{
+			Status: status.Convert(trail.ToGRPC(toSend)).Proto(),
+		})
+		if err != nil {
+			return
+		}
+		st.SetWriteDeadline(time.Now().Add(10 * time.Second))
+		if _, err := st.Write(binary.LittleEndian.AppendUint32(nil, uint32(len(errBuf)))); err != nil {
+			return
+		}
+		if _, err := st.Write(errBuf); err != nil {
+			return
+		}
+		if err := st.Close(); err != nil {
+			return
+		}
+	}
+
+	st.SetReadDeadline(time.Now().Add(10 * time.Second))
+	var reqLen uint32
+	if err := binary.Read(st, binary.LittleEndian, &reqLen); err != nil {
+		log.DebugContext(c.Context(), "Failed to read request size.", "error", err)
+		return
+	}
+	if reqLen >= quicMaxMessageSize {
+		log.WarnContext(c.Context(), "Received oversized request.", "request_len", reqLen)
```

Comment:
Is this warning in any way actionable for someone reading the logs?

Reply:
It would show the misbehaving peer as part of the log fields, but it's either a malicious client or a bug.